sdk/storage/azure-storage-blob/azure/storage/blob/_shared/streams.py

```diff
@@ -27,6 +31,235 @@ class StructuredMessageProperties(IntFlag):
+            else:
+                self._current_region_content = b''
+
+        elif region == SMRegion.MESSAGE_FOOTER:
+            if StructuredMessageProperties.CRC64 in self.flags:
+                self._current_region_content = (
+                    self._message_crc64.to_bytes(StructuredMessageConstants.CRC64_LENGTH, 'little'))
+            else:
+                self._current_region_content = b''
+
+        else:
+            raise StructuredMessageError(f"Invalid metadata SMRegion {self._current_region}")
+
+    def _advance_region(self, current: SMRegion):
+        self._current_region_offset = 0
+        self._current_region_content = b''
+
+        if current == SMRegion.MESSAGE_HEADER:
+            self._current_region = SMRegion.SEGMENT_HEADER
+            self._current_region_length = self._segment_header_length
+        elif current == SMRegion.SEGMENT_HEADER:
+            self._current_region = SMRegion.SEGMENT_CONTENT
+            self._current_region_length = min(self._segment_size, self.content_length - self._content_offset)
+            self._segment_crc64 = 0
+        elif current == SMRegion.SEGMENT_CONTENT:
+            self._current_region = SMRegion.SEGMENT_FOOTER
+            self._current_region_length = self._segment_footer_length
+        elif current == SMRegion.SEGMENT_FOOTER:
+            # If we're at the end of the content
+            if self._content_offset == self.content_length:
+                self._current_region = SMRegion.MESSAGE_FOOTER
+                self._current_region_length = self._message_footer_length
+            else:
+                self._current_region = SMRegion.SEGMENT_HEADER
+                self._current_region_length = self._segment_header_length
+        else:
+            raise StructuredMessageError(f"Invalid SMRegion {self._current_region}")
+
+    def _read_metadata_region(self, region: SMRegion, size: int, output: BytesIO) -> int:
+        # If nothing has been read, buffer the region
+        if self._current_region_offset == 0:
+            self._buffer_metadata_region(region)
+
+        read_size = min(size, self._current_region_length - self._current_region_offset)
+        content = self._current_region_content[self._current_region_offset: self._current_region_offset + read_size]
+        output.write(content)
+
+        self._current_region_offset += read_size
+        if (self._current_region_offset == self._current_region_length and
+                self._current_region != SMRegion.MESSAGE_FOOTER):
+            self._advance_region(region)
+
+        return read_size
+
+    def _read_content(self, size: int, output: BytesIO) -> int:
+        read_size = min(size, self._current_region_length - self._current_region_offset)
+        content = self._inner_stream.read(read_size)
+        output.write(content)
+
+        self._segment_crc64 = crc64.compute_crc64(content, self._segment_crc64)
+        self._message_crc64 = crc64.compute_crc64(content, self._message_crc64)
```

Comment:
interesting you just calculate this no matter what. I suppose it's the only feature currently available and we realistically wouldn't use this encoding otherwise.

Reply:
Yeah, that was probably just an oversight more than anything. We probably shouldn't be calculating these if CRC is not enabled. I'll change it.